arc_cache.go

```diff
@@ -45,31 +53,73 @@ func newARCCachedBS(ctx context.Context, bs Blockstore, lruSize int) (*arccache,
 	return c, nil
 }
 
-func (b *arccache) getLock(k cid.Cid) *sync.RWMutex {
-	return &b.lks[mutexKey(k)]
+func (b *arccache) refLock(k string) *lock {
+	b.lklk.Lock()
+	lk, ok := b.lks[k]
+	if !ok {
+		lk = new(lock)
+		b.lks[k] = lk
+	}
+	lk.refcnt++
+	b.lklk.Unlock()
+	return lk
+}
+
+func (b *arccache) lock(k string) {
+	lk := b.refLock(k)
+	lk.mu.Lock()
+}
+
+func (b *arccache) rlock(k string) {
+	lk := b.refLock(k)
+	lk.mu.RLock()
+}
+
+func (b *arccache) decLock(key string) *lock {
+	b.lklk.Lock()
+	lk := b.lks[key]
+	lk.refcnt--
+	if lk.refcnt == 0 {
+		delete(b.lks, key)
+	}
+	b.lklk.Unlock()
+	return lk
+}
+
+func (b *arccache) unlock(key string) {
+	lk := b.decLock(key)
+	lk.mu.Unlock()
+}
+
+func (b *arccache) runlock(key string) {
+	lk := b.decLock(key)
+	lk.mu.RUnlock()
 }
 
-func mutexKey(k cid.Cid) uint8 {
-	return k.KeyString()[len(k.KeyString())-1]
+func cacheKey(k cid.Cid) string {
+	return string(k.Hash())
 }
 
 func (b *arccache) DeleteBlock(k cid.Cid) error {
 	if !k.Defined() {
 		return nil
 	}
 
-	if has, _, ok := b.queryCache(k); ok && !has {
+	key := cacheKey(k)
```

Comment:
I'm now reusing this key to avoid having to compute/allocate it twice. That's probably why I'm seeing equivalent perf with the striped version.